### PR TITLE
[ODPR-1556] - Add a mechanism to allow for saveing Enrolment data in …

### PR DESCRIPTION
**Pull request by georgievh**
…case first attempt fails.

### CODE REVIEW

File: app/connectors/PendingEnrolmentConnector.scala [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+@Singleton
+class PendingEnrolmentConnector @Inject()(appConfig: FrontendAppConfig, httpClient: HttpClientV2)
+                                         (implicit ec: ExecutionContext) {
+
+  def getPendingEnrolment()(implicit hc: HeaderCarrier): Future[PendingEnrolment] =
+    httpClient.get(url"${appConfig.digitalPlatformReportingUrl}/digital-platform-reporting/pending-enrolment")
+      .execute[HttpResponse]
+      .map { response =>
+        response.status match {
+          case OK => response.json.as[PendingEnrolment]
```
[the remark this answers is not in the document]
Comment (georgievh) [reply app/connectors/PendingEnrolmentConnector.scala:38]: Fixed

File: app/connectors/PendingEnrolmentConnector.scala [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+        }
+      }
+
+  def remove()(implicit hc: HeaderCarrier): Future[Done] =
+    httpClient.delete(url"${appConfig.digitalPlatformReportingUrl}/digital-platform-reporting/pending-enrolment")
+      .execute[HttpResponse]
+      .flatMap { response =>
+        response.status match {
+          case OK => Future.successful(Done)
```
[the remark this answers is not in the document]
Comment (georgievh) [reply app/connectors/PendingEnrolmentConnector.scala:47]: Fixed

File: app/connectors/TaxEnrolmentConnector.scala [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+            (implicit hc: HeaderCarrier): Future[Done] =
+    httpClient.put(url"${appConfig.taxEnrolmentsBaseUrl}/tax-enrolments/enrolments/${upsertKnownFacts.enrolmentKey}")
+      .setHeader("X-Correlation-ID" -> uuidService.generate())
+      .setHeader("X-Conversation-ID" -> uuidService.generate())
+      .setHeader("X-Forwarded-Host" -> appConfig.appName)
+      .withBody(Json.toJson(upsertKnownFacts))
+      .execute[HttpResponse]
+      .flatMap { response =>
+        response.status match {
+          case NO_CONTENT => Future.successful(Done)
```
[the remark this answers is not in the document]
Comment (georgievh) [reply app/connectors/TaxEnrolmentConnector.scala:48]: fixed

File: app/connectors/TaxEnrolmentConnector.scala [hunk 4]
```diff
@@ -0,0 +1,65 @@
...
+                              (implicit hc: HeaderCarrier): Future[Done] =
+    httpClient.post(url"${appConfig.taxEnrolmentsBaseUrl}/tax-enrolments/groups/${groupEnrolment.groupId}/enrolments/${groupEnrolment.enrolmentKey}")
+      .setHeader("X-Correlation-ID" -> uuidService.generate())
+      .setHeader("X-Conversation-ID" -> uuidService.generate())
+      .setHeader("X-Forwarded-Host" -> appConfig.appName)
+      .withBody(Json.toJson(groupEnrolment))
+      .execute[HttpResponse]
+      .flatMap { response =>
+        response.status match {
+          case CREATED | CONFLICT => Future.successful(Done)
```
[the remark this answers is not in the document]
Comment (georgievh) [reply app/connectors/TaxEnrolmentConnector.scala:62]: fixed

File: app/controllers/actions/IdentifierAction.scala [hunk 5]
```diff
@@ -51,16 +54,27 @@ class AuthenticatedIdentifierAction @Inject()(
           Future.successful(Redirect(routes.UnauthorisedController.onPageLoad()))
         }
 
-      case _ =>
-        Future.successful(Redirect(routes.UnauthorisedController.onPageLoad()))
-    } recover {
+      case _ => reEnrol(request, block)(hc).recover {
+        case _ => Redirect(routes.UnauthorisedController.onPageLoad())
```
[the remark this answers is not in the document]
Comment (georgievh) [reply app/controllers/actions/IdentifierAction.scala:58]: fixed

File: app/controllers/actions/IdentifierAction.scala [hunk 6]
```diff
@@ -51,16 +54,27 @@ class AuthenticatedIdentifierAction @Inject()(
...
+      }
+    } recoverWith {
       case _: NoActiveSession =>
-        Redirect(config.loginUrl, Map("continue" -> Seq(config.loginContinueUrl)))
-      case _: AuthorisationException =>
-        Redirect(routes.UnauthorisedController.onPageLoad())
+        Future.successful(Redirect(config.loginUrl, Map("continue" -> Seq(config.loginContinueUrl))))
+      case _: AuthorisationException => reEnrol(request, block)(hc).recover {
+        case _ => Redirect(routes.UnauthorisedController.onPageLoad())
```
[the remark this answers is not in the document]
Comment (georgievh) [reply app/controllers/actions/IdentifierAction.scala:64]: fixed

File: it/test/connectors/TaxEnrolmentConnectorSpec.scala [hunk 7]
```diff
@@ -0,0 +1,78 @@
...
+
+class TaxEnrolmentConnectorSpec extends AnyFreeSpec
+  with Matchers
+  with ScalaFutures
+  with IntegrationPatience
+  with WireMockSupport {
+
+  implicit private val hc: HeaderCarrier = HeaderCarrier()
+
+  private lazy val app: Application = new GuiceApplicationBuilder()
```
[the remark this answers is not in the document]
Comment (georgievh) [reply it/test/connectors/TaxEnrolmentConnectorSpec.scala:38]: Done

File: it/test/connectors/TaxEnrolmentConnectorSpec.scala [hunk 8]
```diff
@@ -0,0 +1,78 @@
...
+    .configure("microservice.services.tax-enrolments.port" -> wireMockPort)
+    .build()
+
+  private lazy val underTest = app.injector.instanceOf[TaxEnrolmentConnector]
+
+  ".allocateEnrolmentToGroup" - {
```
[the remark this answers is not in the document]
Comment (georgievh) [reply it/test/connectors/TaxEnrolmentConnectorSpec.scala:44]: Done

**Review by wolfendale — APPROVED** [review]

**Pull request merged**